File: src/leetcode/array/java/MeetingRoomsII.java
package leetcode.array.java;

import java.util.Arrays;
/*
    time: O(nLogn)
    space: O(n)
 */
public class MeetingRoomsII {
    public int minMeetingRooms(int[][] intervals) {

        int len = intervals.length; //notice the length
        int start[] = new int[len];
        int end[] = new int[len];

        for (int i = 0; i < len; i++) {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }

        Arrays.sort(start);
        Arrays.sort(end);

        int endIdx = 0;
        int res = 0;
        for (int i = 0; i < len; i++) {
            if (start[i] < end[endIdx]) {
                res++;
            } else {
                endIdx++;
            }
        }
        return res;
    }
}

File: src/leetcode/array/kotlin/MeetingRoomsII.kt
package leetcode.array.kotlin

import java.util.*

// do sort
// si < e, rooms++
// else e+
fun minMeetingRooms(intervals: Array<IntArray>): Int {
    val len = intervals.size
    var start = IntArray(len)
    var end = IntArray(len)

    for (i in 0 until len) {
        start[i] = intervals[i][0]
        end[i] = intervals[i][1]
    }
    Arrays.sort(start)
    Arrays.sort(end)
    var res = 0;
    var endIdx = 0;
    start.forEach { s ->
        if (s < end[endIdx]) {
            res++;
        } else {
            endIdx++
        }
    }
    return res;
}